Move Paths into Trees

jalanb opened the pull request:
`paths.py` is too big, it should be split into a number of modules. e.g. 1 module per class, in a new package.

THis branch adds a package, `pysyte.types.trees` as that new package. The name "trees" is temporary, allowing us to keep the old module (`paths.py`) while changes are underway. After all removed from `paths.py` we can re-name the `trees` package back to `paths`.

We should initially split the code as mentioned: one class per module, so e.g. the `FilePath` class might go in a `trees/file.py` (or maybe a `trees/files.py`)

Once that's working we can review the structure for further partitioning of modules, or further refactoring of the ones we have

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

Refactored the `paths.py` module by splitting it into multiple smaller modules under the new package `pysyte.types.trees`. Updated various imports and references across the codebase to use the new package. Enhanced multiple modules by adding new functions, improving the structure, and refining the existing logic.

- **Enhancements**:
    - Refactored the `paths.py` module by splitting it into multiple smaller modules under the new package `pysyte.types.trees`.
    - Introduced new classes and functions in the `pysyte.types.trees` package to handle paths, directories, and files more modularly.
    - Updated various imports and references across the codebase to use the new `pysyte.types.trees` package instead of the old `paths.py` module.
    - Improved the structure and readability of the `methods.py` module by adding docstrings and reorganizing the code.
    - Enhanced the `config/types.py` module by introducing abstract base classes and improving the configuration file handling.
    - Refined the `dictionaries.py` module by adding type annotations and improving the structure of the classes.
    - Updated the `getch.py` module to include new functions for handling key names and added a main function for script execution.
    - Improved the `streams.py` module by adding new functions for printing to stdout and stderr, and enhancing the context managers for swallowing output.
    - Enhanced the `arguments.py` module by refining the argument parsing logic and improving the structure of the `ArgumentsParser` class.
    - Refined the `splits.py` module by adding type annotations and improving the string splitting and joining functions.
    - Improved the `strings.py` module by adding new utility functions and enhancing the existing classes.
    - Enhanced the `lots.py` module by adding new functions and improving the structure of the `Many` and `Lots` classes.
    - Refined the `paths.py` module by adding new functions for handling paths and improving the structure of the existing functions.
    - Improved the `methods.py` module by adding new utility functions and enhancing the existing classes.
    - Enhanced the `lines.py` module by adding new functions for handling lines and improving the structure of the existing functions.
    - Refined the `importers.py` module by adding new functions for handling imports and improving the structure of the existing functions.
    - Improved the `platforms.py` module by refining the platform handling logic and improving the structure of the existing functions.
    - Enhanced the `similarities.py` module by adding new functions for calculating similarities and improving the structure of the existing functions.
    - Refined the `test_lines.py` module by adding new test cases and improving the structure of the existing tests.
    - Improved the `xdg.py` module by refining the XDG configuration handling logic and improving the structure of the existing functions.

<!-- Generated by sourcery-ai[bot]: end summary -->

jalanb committed:
Prefer composition over inheritance

jalanb committed:
Clarify code

Better names, add docstrings and type hints

(with help from chatgpt)

jalanb committed:
Move file types to their own module

jalanb committed:
Blacken code

jalanb committed:
Clarify code

jalanb committed:
Blacken code

jalanb committed:
Flaken code

jalanb committed:
Are we there yet?

jalanb committed:
move paths into trees

jalanb committed:
Add function to find calling shell

jalanb committed:
Add more stream handling

Add print_out() and print_err() for consistency in code
Add quiet* context handlers to turn streams off

jalanb committed:
Add pysyte.types.time.durations module

jalanb committed:
Exapand types.lists into a package

Add types.lists.search

jalanb committed:
Add more typing info

jalanb committed:
Caculate better names for options

jalanb committed:
Prefer name "paths"

jalanb committed:
Make pysyte/oss/getch.py runnable

jalanb committed:
Be more accurate

jalanb committed:
Expand the durations class

jalanb committed:
Require lazy and textual

jalanb committed:
An abstract base class for configuration files

jalanb committed:
Add more typing info

jalanb committed:
Better naming

jalanb committed:
Better naming

jalanb committed:
Add red_def() to read a fundtion definition

jalanb committed:
Add explanation to doctest

jalanb committed:
Test pysyte.types.methods

jalanb committed:
Add doctests for pysyte.types.time.times

jalanb committed:
Allow adding ext to a Filepath

jalanb committed:
move paths into trees

jalanb committed:
Better naming

jalanb committed:
Use new packages

jalanb committed:
Allow adding string to DirPath

jalanb committed:
Extract a RealPath class, to handle real paths

jalanb committed:
Add methods to RealPath

jalanb committed:
Add Class for Chmod Values

jalanb committed:
Add a separator to StringPath

Allows better adding/splitting

jalanb committed:
Add pp() for quick debugging

jalanb committed:
Extract a Repper class

jalanb committed:
Blacken code

jalanb committed:
Flaken code

jalanb committed:
Untypo NonePath -> NoPath

jalanb committed:
gather all makes into the makes module

jalanb committed:
Move "path" aliases into paths module

jalanb commented:
@CodiumAI-Agent /review

jalanb commented:
@sourcery-ai review

QodoAI-Agent commented:
## PR Reviewer Guide 🔍

#### (Review updated until commit https://github.com/jalanb/pysyte/commit/f8fd5cdb6d6e763ada38d786ba55064156f6e689)


Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<a href='https://github.com/jalanb/pysyte/pull/72/files#diff-ec1667fbc1139b96f34413afa4d10a9d09f21a62deadf9133d7faf192df5dbbfR34-R38'><strong>Possible Bug</strong></a><br>The method `isroot` in `RealPath` class may not handle cases correctly where the path is not at the root but still returns true due to string comparison logic.

<a href='https://github.com/jalanb/pysyte/pull/72/files#diff-7c30d8cb949cd89931cf05ddeca91c67b851afce757ce6e2851eee0268179f87R24-R26'><strong>Redundant Code</strong></a><br>The `__add__` method in `FilePath` class seems redundant as it only calls another method without additional logic.

<a href='https://github.com/jalanb/pysyte/pull/72/files#diff-07564e04808aba30d8cd69ab4d280887da3f18606a91a4ae965826c59745a6adR88-R92'><strong>Performance Issue</strong></a><br>The method `list_dirs_files` might be inefficient due to multiple list comprehensions inside a loop which could be optimized.

</td></tr>
</table>

jalanb committed:
Add a std printer to streams

jalanb committed:
Expand std

jalanb committed:
Add more imports to test

jalanb committed:
Add StringFile, to let strings look like files

jalanb committed:
Move chmod names to own module

jalanb committed:
Make ExtendedPath ako FilePath

jalanb committed:
Add Extension Handling to ExtendedPath and Perform Code Cleanups

- Implemented extension handling methods in ExtendedPath, enhancing its functionality.
- Refactored StringFile to support multiple string arguments during initialization.
- Ensured consistent return types across methods, adhering to the package's API standards.
- Improved path handling logic in makes.py for clarity and efficiency.
- Added __post_init__ to RealPath for additional initialization logic.
- Cleaned up import statements for better readability and maintainability.

jalanb committed:
Move fnmatch funcs to own module

jalanb committed:
Blacken code

jalanb committed:
Acccept Flake8 suggestions

jalanb committed:
Appreciate flake8/mypy

QodoAI-Agent commented:
**[Persistent review](https://github.com/jalanb/pysyte/pull/72#issuecomment-2236830446)** updated to latest commit https://github.com/jalanb/pysyte/commit/f8fd5cdb6d6e763ada38d786ba55064156f6e689
